### PR TITLE
improved uuid checks on commit hook and make targets

### PR DESCRIPTION
**Pull request by rochacbruno**
Following recommendations from our meeting.

- Put the fix uuids scripts all together in a single file
- Now there is only `make uuid-check` and `make uuid-fix`
- `make install-commit-hook` includes a call to `uuid-fix` to fix id issues also `can-i-push?` to check if everyhting is ok
- added check for `@id:xyz` missing whitespace after : as in `@id: xyz`
- flake8 will be checked only for modified files 
- travis will run `check-uuid` again once PR is sent

TODO:
- Automatically include uuid on all tests which misses `@id` (currently we need to keep an empty one to be filled) 
- Move all that logic to testimony itself as it is already checking for tags issues and there is issue opened there. https://github.com/SatelliteQE/testimony/issues/121

### CODE REVIEW

**Comment by oshtaier** [comment]
ACK. Missed all that shell magic, but as far as I see two files were just merged into one

**Comment by rochacbruno** [comment]
@oshtaier regex + grep + sed = magic

File: scripts/check_uuids.sh [hunk 1]
```diff
@@ -3,19 +3,21 @@
 # This script checks duplicated or empty uuids and exit with 1 if found
 
 # finds occurrences of empty @id: testimony tags
-grep -E -i -r -n "@id:(.+[[:blank:]]|$)" tests/foreman/
+grep -E -i -r -n "@id:(.+[[:blank:]]|$)" tests/foreman/ --include=*.py
```
Comment (omaciel) [inline scripts/check_uuids.sh]: 👍

File: scripts/check_uuids.sh [hunk 2]
```diff
@@ -3,19 +3,21 @@
...
+else
+    echo "No empty @id found"
 fi
 
 # Finds occurrences of @id: in testimony tags then
 # sort the output and filters only the duplicated
 # then looks for existence of "@id:" in final output
 # NOTE: can't print the line number -n here because of uniq -d
-grep -r -i "@id:" tests/foreman/ | sort | uniq -d | grep "@id:"
+grep -r -i "@id:" tests/foreman/ --include=*.py | sort | uniq -d | grep "@id:"
```
Comment (omaciel) [inline scripts/check_uuids.sh]: 👍

File: scripts/fix_uuids.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+# Then it finds for duplicates and replaces last occurrence with new uuids
+# finally it fixes '@id:xyz' to '@id: xyz' (missing space after :)
+
+# finds occurrences of empty @id: testimony tags
+EMPTY_IDS=$(grep -E -i -r -n "@id:(.+[[:blank:]]|$)" tests/foreman/ --include=*.py)
+
+if [ -n "$EMPTY_IDS" ]; then
+   echo "Generating new UUIDS for empty @id tags..."
+else
+   echo "No empty @id was found"
```
Comment (omaciel) [inline scripts/fix_uuids.sh]: Do you need to explicitly exit here?

File: scripts/fix_uuids.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+fi
+
+# iterate if any empty @id found
+for output_line in $EMPTY_IDS
+do
+    if (echo "$output_line" | grep "tests/foreman"); then
```
Comment (omaciel) [inline scripts/fix_uuids.sh]: Is this check required? Your grep is already pointing at `tests/foreman` so I expect that all lines would have this?

File: scripts/fix_uuids.sh (hunk 3 above)
Comment (rochacbruno) [inline scripts/fix_uuids.sh]: No need to exit here because I want the other checks in script to work until the end of the script.

File: scripts/fix_uuids.sh (hunk 4 above)
Comment (rochacbruno) [inline scripts/fix_uuids.sh]: I needed to include it here because in some cases grep output have been wrapped, but I can try different approach.

File: Makefile [hunk 5]
```diff
@@ -132,26 +131,24 @@ logs-join:
...
-
-uuid-replace-empty:  ## list duplicated uuids
-	scripts/replace_empty_uuids.sh
+uuid-fix:
+	@scripts/fix_uuids.sh
 
 flake8:
 	$(info "Checking style and syntax errors with flake8 linter...")
-	@flake8 . --show-source
+	@flake8 $(shell git diff --name-only) --show-source
```
Comment (elyezer) [inline Makefile]: This is a good change, but I feel that it is more suitable for travis and commit hook other than checking local changes, I mean I will not be able to run `make flake8` for files that I have not added to git yet.

File: scripts/check_uuids.sh [hunk 6]
```diff
@@ -24,4 +26,18 @@ DUPLICATE=$?
...
+# finds missing empty space after @id:
+grep -E -i -r -n "@id:[^ ]" tests/foreman/ --include=*.py
+MISSING_SPACES=$?
+
+if [ $MISSING_SPACES = 0 ]; then
+   echo "Found @id: tags missing space after : ..."
+   exit 1
+else
+   echo "No @id missing spaces after : was found"
+fi
```
Comment (elyezer) [inline scripts/check_uuids.sh]: Missing new line at the end of this line.

**Comment by elyezer** [comment]
My comments are not a blocker, also don't wait for me to get this merged.

File: Makefile (hunk 5 above)
Comment (rochacbruno) [inline Makefile]: @elyezer should I change the target name to `gitflake` or add a new target?

Comment (elyezer) [inline Makefile]: I mean, that is a personal preference but if others don't worry about it then it is fine to leave instead of creating a separated target.

I've just pointed some of the usecase I personally have.

File: scripts/check_uuids.sh (hunk 6 above)
Comment (rochacbruno) [inline scripts/check_uuids.sh]: done!

File: Makefile (hunk 5 above)
Comment (rochacbruno) [inline Makefile]: changed to `gitflake8`

**Comment by rochacbruno** [comment]
@omaciel token changed to `$ID_TOKEN` constant

@elyezer your comments addressed

**Comment by elyezer** [comment]
Thank you @rochacbruno, will leave to @omaciel have a second look before merging.

**Comment by omaciel** [comment]
@rochacbruno 

```console
2016-09-27 11:47:39 < omaciel> brunorocha|brb: about https://github.com/SatelliteQE/robottelo/pull/3855 I see that 
                               $ID_TOKEN is defined in 2 scripts. Perhaps it could be defined in one place and 
                               sourced in the other 2 scripts?
```

**Comment by rochacbruno** [comment]
@omaciel fixed! I merged the solution in a single script `fix_uuids.sh` which now accepts a `--check` argument to run in check only mode. So we have now the `$ID_TOKEN` defined in a single place and a single script to maintain.
